File: 323. Number of Connected Components in an Undirected Graph.cpp
class Solution {
public:
    // DFS: O(N2)time & space, 42.76%, ok
    int countComponents(int n, vector<pair<int, int>>& edges) {
        int ret = 0, i = 0;
        if(edges.empty()) return n;  // bug: individual points are connected components too
        vector<int> visited(n, 0);
        unordered_map<int, vector<int>> adj;
        for(int j = 0; j < n; j++) adj[j] = vector<int>();
        for(auto p: edges){
            adj[p.first].push_back(p.second);
            adj[p.second].push_back(p.first);
        }
        while(i < n) {
            if(visited[i]) {i++; continue;}
            dfs(i, adj, ret, visited);
            i++; ret++;
        }
        return ret;
    }

    void dfs(int i, unordered_map<int, vector<int>>& adj, int& ret, vector<int>& visited){
        if(visited[i]) return;
        visited[i] = 1;
        for(auto node: adj[i]) dfs(node, adj, ret, visited);
    }

    // union find: O(mlogN)time (m = vectorSize), O(N)space, 42.7%, ok
    int countComponents(int n, vector<pair<int, int>>& edges) {
        int ret = n;
        vector<int> root;
        for(int i = 0; i < n; i++) root.push_back(i);
        for(auto e: edges){
            int root1, root2;
            int i = e.first, j = e.second;
            while(i != root[i]) i = root[i];
            while(j != root[j]) j = root[j];
            if(i != j){
                root[i] = j;
                ret--;
            }
        }
        return ret;
    }
};

File: 323. Number of Connected Components in an Undirected Graph.py
class Solution(object):
    
    # dfs: O(E) time, O(E) space
    def countComponents(self, n, edges):
        def dfs(node):
            if node in self.adj:
                neighbors = self.adj[node]
                del self.adj[node]
                for n in neighbors:
                    dfs(n)
        ret = 0
        self.adj = {i: [i] for i in xrange(n)}
        for a, b in edges:
            self.adj[a] += [b]
            self.adj[b] += [a]  # bug: undirected graph
            
        for a, b in edges:
            if a in self.adj:  # check visited
                ret += 1
                dfs(a)
        return ret + len(self.adj)
    
    
    # union find: O(V + E) time, O(V) space
    def countComponents(self, n, edges):
        """
        :type n: int
        :type edges: List[List[int]]
        :rtype: int
        """
        def find(node):
            if self.root[node] == node:
                return node
            self.root[node] = find(self.root[node])
            return self.root[node]
        
        self.root = {i:i for i in xrange(n)}
        ret = n
        for a, b in edges:
            root_a = find(a)
            root_b = find(b)
            if root_a != root_b:
                self.root[root_a] = root_b   # bug: root[a] = b
                ret -= 1
        return ret
        